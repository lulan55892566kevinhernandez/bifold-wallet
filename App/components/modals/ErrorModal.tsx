import React, { useContext, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Dimensions, Modal, StyleSheet } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'

import { Context } from '../../store/Store'
import { ColorPallet } from '../../theme'

import BifoldErrorX from './BifoldErrorX'

const { height } = Dimensions.get('window')
const genericErrorCode = 1024
const styles = StyleSheet.create({
  container: {
    minHeight: height,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: ColorPallet.brand.primary,
  },
})

const ErrorModal: React.FC = () => {
  const { t } = useTranslation()
  const [modalVisible, setModalVisible] = useState<boolean>(false)
  const [state] = useContext(Context)
  const onDismissModalTouched = () => {
    setModalVisible(false)
  }

  useEffect(() => {
    if (state.error && state.error.title && state.error.message && state.error.code) {
      setModalVisible(true)

      return
    }

    setModalVisible(true)
  }, [state])

  return (
    <Modal visible={modalVisible} transparent={true}>
      <SafeAreaView style={[styles.container]}>
        <BifoldErrorX
          title={state.error ? state.error.title : t('Error.Unknown')}
          message={state.error ? state.error.message : t('Error.Problem')}
          code={state.error ? state.error.code : genericErrorCode}
          onPress={onDismissModalTouched}
        />
      </SafeAreaView>
    </Modal>
  )
}

export default ErrorModal
