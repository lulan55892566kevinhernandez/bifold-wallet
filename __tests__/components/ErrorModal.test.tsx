import { render, waitFor, fireEvent } from '@testing-library/react-native'
import React, { useContext } from 'react'

import ErrorModal from '../../App/components/modals/ErrorModal'

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn(),
}))

const setState = jest.fn()
const state = {
  onboarding: {
    DidAgreeToTerms: false,
    DidCompleteTutorial: false,
    DidCreatePIN: false,
  },
  notifications: {
    ConnectionPending: true,
  },
  error: null,
}

// @ts-ignore
useContext.mockImplementation(() => [state, setState])

describe('ErrorModal Component', () => {
  test('Renders correctly', async () => {
    const tree = render(<ErrorModal />)

    expect(tree).toMatchSnapshot()
    expect(tree.queryByText('Global.Okay')).not.toBeNull()
  })

  test('Dismiss on demand', async () => {
    const tree = render(<ErrorModal />)

    const dismissBtn = await tree.findByText('Global.Okay')
    fireEvent(dismissBtn, 'press')

    expect(tree).toMatchSnapshot()
  })
})
