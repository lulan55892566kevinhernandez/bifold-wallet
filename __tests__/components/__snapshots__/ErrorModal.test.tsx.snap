// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorModal Component Dismiss on demand 1`] = `
<Modal
  hardwareAccelerated={false}
  transparent={true}
  visible={false}
>
  <RNCSafeAreaView
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#42803E",
          "flexDirection": "column",
          "justifyContent": "center",
          "minHeight": 1334,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#313132",
          "borderColor": "#D8292F",
          "borderRadius": 5,
          "borderWidth": 1,
          "marginHorizontal": 25,
          "minWidth": 700,
          "padding": 10,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "paddingHorizontal": 5,
            "paddingTop": 5,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignSelf": "center",
                "marginRight": 10,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#0099FF",
                  "fontSize": 30,
                },
                undefined,
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
        <Text
          style={
            Object {
              "alignSelf": "center",
              "color": "#FFFFFF",
              "flexShrink": 1,
              "fontSize": 18,
              "fontWeight": "bold",
            }
          }
        >
          Error.Unknown
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "column",
            "marginLeft": 40,
            "paddingBottom": 5,
            "paddingHorizontal": 5,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "flexShrink": 1,
              "fontSize": 18,
              "fontWeight": "normal",
              "marginVertical": 15,
              "paddingBottom": 10,
            }
          }
        >
          Error.Problem
        </Text>
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "flexShrink": 1,
              "fontSize": 18,
              "fontWeight": "normal",
              "marginVertical": 15,
              "paddingBottom": 10,
            }
          }
        >
          Global.ErrorCode 1024
        </Text>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={false}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#35823F",
              "borderRadius": 4,
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 18,
                  "fontWeight": "bold",
                  "textAlign": "center",
                },
                false,
              ]
            }
          >
            Global.Okay
          </Text>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</Modal>
`;

exports[`ErrorModal Component Renders correctly 1`] = `
<Modal
  hardwareAccelerated={false}
  transparent={true}
  visible={false}
>
  <RNCSafeAreaView
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#42803E",
          "flexDirection": "column",
          "justifyContent": "center",
          "minHeight": 1334,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#313132",
          "borderColor": "#D8292F",
          "borderRadius": 5,
          "borderWidth": 1,
          "marginHorizontal": 25,
          "minWidth": 700,
          "padding": 10,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "paddingHorizontal": 5,
            "paddingTop": 5,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignSelf": "center",
                "marginRight": 10,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#0099FF",
                  "fontSize": 30,
                },
                undefined,
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
        <Text
          style={
            Object {
              "alignSelf": "center",
              "color": "#FFFFFF",
              "flexShrink": 1,
              "fontSize": 18,
              "fontWeight": "bold",
            }
          }
        >
          Error.Unknown
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "column",
            "marginLeft": 40,
            "paddingBottom": 5,
            "paddingHorizontal": 5,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "flexShrink": 1,
              "fontSize": 18,
              "fontWeight": "normal",
              "marginVertical": 15,
              "paddingBottom": 10,
            }
          }
        >
          Error.Problem
        </Text>
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "flexShrink": 1,
              "fontSize": 18,
              "fontWeight": "normal",
              "marginVertical": 15,
              "paddingBottom": 10,
            }
          }
        >
          Global.ErrorCode 1024
        </Text>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={false}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#35823F",
              "borderRadius": 4,
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 18,
                  "fontWeight": "bold",
                  "textAlign": "center",
                },
                false,
              ]
            }
          >
            Global.Okay
          </Text>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</Modal>
`;
